sui-execution/latest/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -521,22 +499,26 @@ mod checked {
             previous_package: &MovePackage,
             new_modules: &[CompiledModule],
             dependencies: impl IntoIterator<Item = &'p MovePackage>,
-        ) -> Result<Object, ExecutionError> {
-            Object::new_upgraded_package(
-                previous_package,
+        ) -> Result<MovePackage, ExecutionError> {
+            previous_package.new_upgraded(
                 storage_id,
                 new_modules,
-                self.tx_context.digest(),
                 self.protocol_config,
                 dependencies,
             )
         }
 
         /// Add a newly created package to write as an effect of the transaction
-        pub fn write_package(&mut self, package: Object) -> Result<(), ExecutionError> {
-            assert_invariant!(package.is_package(), "Must be a package");
+        pub fn write_package(&mut self, package: MovePackage) {
             self.new_packages.push(package);
-            Ok(())
+        }
+
+        /// Return the last pacakge pushed in `write_package`.
+        /// This function should be used in block of codes that push a pacakge, verify
+        /// it, run the init and in case of error will remove the package.
+        /// The package has to be pushed for the init to run correctly.
```

Comment:
you can check the caller to see what this means. 
Not pretty and happy to get feedback if someone feels strong about it, but I think it's a fine step for now.
Once we plug in the loader in a more structured and reasonable way this may all go away